File: agents/services/a2i-service.ts
/**
 * Amazon Augmented AI (A2I) Service
 *
 * Handles human-in-the-loop review escalation for high-risk contracts.
 */

import {
  SageMakerA2IRuntimeClient,
  StartHumanLoopCommand,
  StopHumanLoopCommand,
  DescribeHumanLoopCommand,
  ListHumanLoopsCommand,
} from '@aws-sdk/client-sagemaker-a2i-runtime';
import {
  HumanLoopRequest,
  HumanLoopResponse,
} from '../types/risk-scoring.types';

export interface A2IConfig {
  flowDefinitionArn: string;
  region: string;
}

export class A2IService {
  private client: SageMakerA2IRuntimeClient;
  private flowDefinitionArn: string;

  constructor(config: A2IConfig) {
    this.client = new SageMakerA2IRuntimeClient({ region: config.region });
    this.flowDefinitionArn = config.flowDefinitionArn;
  }

  /**
   * Create a human loop for manual review
   */
  async createHumanLoop(request: HumanLoopRequest): Promise<HumanLoopResponse> {
    const command = new StartHumanLoopCommand({
      HumanLoopName: request.humanLoopName,
      FlowDefinitionArn: request.flowDefinitionArn || this.flowDefinitionArn,
      HumanLoopInput: {
        InputContent: JSON.stringify(request.inputContent),
      },
      DataAttributes: request.dataAttributes,
    });

    try {
      const response = await this.client.send(command);
      console.log(`Human loop created: ${response.HumanLoopArn}`);

      return {
        humanLoopArn: response.HumanLoopArn || '',
        humanLoopName: request.humanLoopName,
        status: 'InProgress',
      };
    } catch (error) {
      console.error('Error creating human loop:', error);
      throw new Error(`Failed to create human loop: ${error.message}`);
    }
  }

  /**
   * Get human loop status
   */
  async getHumanLoopStatus(humanLoopName: string): Promise<HumanLoopResponse> {
    const command = new DescribeHumanLoopCommand({
      HumanLoopName: humanLoopName,
    });

    try {
      const response = await this.client.send(command);

      return {
        humanLoopArn: response.HumanLoopArn || '',
        humanLoopName: humanLoopName,
        status: response.HumanLoopStatus as any,
        outputDestination: response.HumanLoopOutput?.OutputS3Uri,
      };
    } catch (error) {
      console.error('Error getting human loop status:', error);
      throw new Error(`Failed to get human loop status: ${error.message}`);
    }
  }

  /**
   * Stop a human loop
   */
  async stopHumanLoop(humanLoopName: string): Promise<void> {
    const command = new StopHumanLoopCommand({
      HumanLoopName: humanLoopName,
    });

    try {
      await this.client.send(command);
      console.log(`Human loop stopped: ${humanLoopName}`);
    } catch (error) {
      console.error('Error stopping human loop:', error);
      throw new Error(`Failed to stop human loop: ${error.message}`);
    }
  }

  /**
   * List human loops by status
   */
  async listHumanLoops(
    flowDefinitionArn?: string,
    createdAfter?: Date,
    maxResults: number = 100
  ): Promise<any[]> {
    const command = new ListHumanLoopsCommand({
      FlowDefinitionArn: flowDefinitionArn || this.flowDefinitionArn,
      CreationTimeAfter: createdAfter,
      MaxResults: maxResults,
      SortOrder: 'Descending',
    });

    try {
      const response = await this.client.send(command);
      return response.HumanLoopSummaries || [];
    } catch (error) {
      console.error('Error listing human loops:', error);
      throw new Error(`Failed to list human loops: ${error.message}`);
    }
  }

  /**
   * Check if human loop is complete
   */
  async isHumanLoopComplete(humanLoopName: string): Promise<boolean> {
    const status = await this.getHumanLoopStatus(humanLoopName);
    return status.status === 'Completed' || status.status === 'Failed' || status.status === 'Stopped';
  }

  /**
   * Wait for human loop completion
   */
  async waitForCompletion(
    humanLoopName: string,
    maxWaitTimeMs: number = 3600000, // 1 hour
    pollIntervalMs: number = 30000 // 30 seconds
  ): Promise<HumanLoopResponse> {
    const startTime = Date.now();

    while (Date.now() - startTime < maxWaitTimeMs) {
      const status = await this.getHumanLoopStatus(humanLoopName);

      if (status.status === 'Completed') {
        console.log(`Human loop completed: ${humanLoopName}`);
        return status;
      } else if (status.status === 'Failed' || status.status === 'Stopped') {
        throw new Error(`Human loop ${status.status.toLowerCase()}: ${humanLoopName}`);
      }

      // Wait before polling again
      await new Promise((resolve) => setTimeout(resolve, pollIntervalMs));
    }

    throw new Error(`Human loop timed out after ${maxWaitTimeMs}ms: ${humanLoopName}`);
  }
}

File: agents-python/services/a2i_service.py
"""
Amazon Augmented AI (A2I) Service

Handles human-in-the-loop review escalation for high-risk contracts.
"""

import boto3
import time
from typing import List, Optional

from types.risk_scoring_types import HumanLoopRequest, HumanLoopResponse


class A2IService:
    """Amazon A2I service for human review escalation"""

    def __init__(self, flow_definition_arn: str, region: str = "us-east-1"):
        """
        Initialize A2I service

        Args:
            flow_definition_arn: A2I flow definition ARN
            region: AWS region
        """
        self.flow_definition_arn = flow_definition_arn
        self.region = region
        self.client = boto3.client('sagemaker-a2i-runtime', region_name=region)

    def create_human_loop(self, request: HumanLoopRequest) -> HumanLoopResponse:
        """
        Create a human loop for manual review

        Args:
            request: Human loop request

        Returns:
            HumanLoopResponse with ARN and status
        """
        try:
            response = self.client.start_human_loop(
                HumanLoopName=request.human_loop_name,
                FlowDefinitionArn=request.flow_definition_arn or self.flow_definition_arn,
                HumanLoopInput={
                    'InputContent': str(request.input_content)
                },
                DataAttributes=request.data_attributes or {}
            )

            print(f"Human loop created: {response.get('HumanLoopArn', '')}")

            return HumanLoopResponse(
                human_loop_arn=response.get('HumanLoopArn', ''),
                human_loop_name=request.human_loop_name,
                status='InProgress'
            )

        except Exception as e:
            print(f"Error creating human loop: {e}")
            raise Exception(f"Failed to create human loop: {str(e)}")

    def get_human_loop_status(self, human_loop_name: str) -> HumanLoopResponse:
        """
        Get human loop status

        Args:
            human_loop_name: Human loop name

        Returns:
            HumanLoopResponse with current status
        """
        try:
            response = self.client.describe_human_loop(
                HumanLoopName=human_loop_name
            )

            return HumanLoopResponse(
                human_loop_arn=response.get('HumanLoopArn', ''),
                human_loop_name=human_loop_name,
                status=response.get('HumanLoopStatus', 'Unknown'),
                output_destination=response.get('HumanLoopOutput', {}).get('OutputS3Uri')
            )

        except Exception as e:
            print(f"Error getting human loop status: {e}")
            raise Exception(f"Failed to get human loop status: {str(e)}")

    def stop_human_loop(self, human_loop_name: str) -> None:
        """
        Stop a human loop

        Args:
            human_loop_name: Human loop name
        """
        try:
            self.client.stop_human_loop(HumanLoopName=human_loop_name)
            print(f"Human loop stopped: {human_loop_name}")
        except Exception as e:
            print(f"Error stopping human loop: {e}")
            raise Exception(f"Failed to stop human loop: {str(e)}")

    def list_human_loops(
        self,
        flow_definition_arn: Optional[str] = None,
        created_after: Optional[str] = None,
        max_results: int = 100
    ) -> List[dict]:
        """
        List human loops by status

        Args:
            flow_definition_arn: Flow definition ARN (optional)
            created_after: Filter by creation date (optional)
            max_results: Maximum number of results

        Returns:
            List of human loop summaries
        """
        try:
            params = {
                'FlowDefinitionArn': flow_definition_arn or self.flow_definition_arn,
                'MaxResults': max_results,
                'SortOrder': 'Descending'
            }

            if created_after:
                params['CreationTimeAfter'] = created_after

            response = self.client.list_human_loops(**params)
            return response.get('HumanLoopSummaries', [])

        except Exception as e:
            print(f"Error listing human loops: {e}")
            raise Exception(f"Failed to list human loops: {str(e)}")

    def is_human_loop_complete(self, human_loop_name: str) -> bool:
        """
        Check if human loop is complete

        Args:
            human_loop_name: Human loop name

        Returns:
            True if complete, False otherwise
        """
        status = self.get_human_loop_status(human_loop_name)
        return status.status in ['Completed', 'Failed', 'Stopped']

    def wait_for_completion(
        self,
        human_loop_name: str,
        max_wait_time_seconds: int = 3600,
        poll_interval_seconds: int = 30
    ) -> HumanLoopResponse:
        """
        Wait for human loop completion

        Args:
            human_loop_name: Human loop name
            max_wait_time_seconds: Maximum wait time (default: 1 hour)
            poll_interval_seconds: Polling interval (default: 30 seconds)

        Returns:
            HumanLoopResponse when complete

        Raises:
            Exception if timeout or loop fails
        """
        start_time = time.time()

        while time.time() - start_time < max_wait_time_seconds:
            status = self.get_human_loop_status(human_loop_name)

            if status.status == 'Completed':
                print(f"Human loop completed: {human_loop_name}")
                return status
            elif status.status in ['Failed', 'Stopped']:
                raise Exception(f"Human loop {status.status.lower()}: {human_loop_name}")

            # Wait before polling again
            time.sleep(poll_interval_seconds)

        raise Exception(f"Human loop timed out after {max_wait_time_seconds}s: {human_loop_name}")
